File: Autonomous Mobile Robots/Chapter9/Lesson4/Chapter9_Lesson4.m
% Chapter9_Lesson4.m
% Elevation + traversability maps (outdoor AMR) — MATLAB + Simulink-friendly script.
% This script:
%   1) Generates a synthetic point cloud for terrain
%   2) Fuses points into a probabilistic elevation grid (per-cell 1D Kalman)
%   3) Computes slope/roughness/step and a traversability cost map
%   4) Optionally builds a simple Simulink model for the per-cell Kalman update

clear; clc; close all;

% -----------------------------
% Parameters
% -----------------------------
xMin = -10; xMax = 10;
yMin = -10; yMax = 10;
res = 0.20;

nx = ceil((xMax - xMin)/res);
ny = ceil((yMax - yMin)/res);

sigma0 = 2.0;       % initial std (m)
measSigma = 0.10;   % sensor std (m)
R = measSigma^2;
gateK = 3.0;

mu = zeros(nx, ny);
sigma2 = (sigma0^2) * ones(nx, ny);
count = zeros(nx, ny);

% -----------------------------
% Synthetic point cloud
% -----------------------------
rng(7);
nGround = 5e4;
nOut = 2e3;

x = (xMin + (xMax-xMin)*rand(nGround,1));
y = (yMin + (yMax-yMin)*rand(nGround,1));

zTrue = 0.20*sin(0.35*x) + 0.15*cos(0.25*y) + 0.03*x ...
        + 0.25*exp(-0.08*((x-2).^2 + (y+1).^2));

zMeas = zTrue + measSigma*randn(nGround,1);
ground = [x, y, zMeas];

xo = (xMin + (xMax-xMin)*rand(nOut,1));
yo = (yMin + (yMax-yMin)*rand(nOut,1));
zo = 0.20*sin(0.35*xo) + 0.15*cos(0.25*yo) + 0.03*xo ...
     + 0.25*exp(-0.08*((xo-2).^2 + (yo+1).^2)) ...
     + 0.8 + 0.05*randn(nOut,1);
outl = [xo, yo, zo];

pts = [ground; outl];
pts = pts(randperm(size(pts,1)), :);

% -----------------------------
% Per-cell Kalman fusion
% -----------------------------
accepted = 0;
for k = 1:size(pts,1)
    px = pts(k,1); py = pts(k,2); pz = pts(k,3);

    ix = floor((px - xMin)/res) + 1;
    iy = floor((py - yMin)/res) + 1;

    if ix < 1 || ix > nx || iy < 1 || iy > ny
        continue;
    end

    m = mu(ix,iy);
    s2 = sigma2(ix,iy);

    nu = pz - m;
    S = s2 + R;

    if count(ix,iy) > 0 && abs(nu) > gateK*sqrt(S)
        continue; % reject outlier/obstacle
    end

    K = s2 / S;
    mu(ix,iy) = m + K*nu;
    sigma2(ix,iy) = (1 - K)*s2;
    count(ix,iy) = count(ix,iy) + 1;
    accepted = accepted + 1;
end

fprintf('Points accepted: %d out of %d\n', accepted, size(pts,1));

% -----------------------------
% Features: slope, roughness, step
% -----------------------------
slope = zeros(nx,ny);
rough = zeros(nx,ny);
stepH = zeros(nx,ny);

for i = 2:nx-1
    for j = 2:ny-1
        if count(i,j) == 0
            continue;
        end

        dzdx = (mu(i+1,j) - mu(i-1,j)) / (2*res);
        dzdy = (mu(i,j+1) - mu(i,j-1)) / (2*res);
        slope(i,j) = atan(sqrt(dzdx^2 + dzdy^2));

        neigh = mu(i-1:i+1, j-1:j+1);
        rough(i,j) = std(neigh(:));

        stepH(i,j) = max(abs(neigh(:) - mu(i,j)));
    end
end

% -----------------------------
% Traversability cost
% -----------------------------
slopeRef = 0.35;  % rad
roughRef = 0.08;  % m
stepRef  = 0.12;  % m

wS = 3.0; wR = 2.0; wD = 2.5; bias = -3.0;

s = slope / max(1e-9, slopeRef);
r = rough / max(1e-9, roughRef);
d = stepH / max(1e-9, stepRef);
score = wS*s + wR*r + wD*d + bias;

cost = 1 ./ (1 + exp(-score));

% -----------------------------
% Plots
% -----------------------------
figure; imagesc(mu'); axis image; colorbar;
title('Elevation mean (m)'); set(gca,'YDir','normal');

figure; imagesc(cost'); axis image; colorbar;
title('Traversability cost (0 easy, 1 hard)'); set(gca,'YDir','normal');

% -----------------------------
% Export
% -----------------------------
writematrix(mu, 'Chapter9_Lesson4_elevation_mu.csv');
writematrix(cost, 'Chapter9_Lesson4_traversability_cost.csv');
writematrix(count, 'Chapter9_Lesson4_fusion_count.csv');

% -----------------------------
% Optional: build a Simulink model that implements ONE cell's 1D Kalman update.
% The idea is to show how the recursion (mu, sigma2) can be implemented as discrete blocks.
% Run: buildSimulinkKalmanCell();
% -----------------------------
% buildSimulinkKalmanCell();

function buildSimulinkKalmanCell()
    model = 'Chapter9_Lesson4_KalmanCell';
    if bdIsLoaded(model); close_system(model, 0); end
    new_system(model); open_system(model);

    add_block('simulink/Sources/In1', [model '/z_meas'], 'Position', [30 40 60 60]);
    add_block('simulink/Sources/In1', [model '/R'], 'Position', [30 90 60 110]);
    add_block('simulink/Discrete/Unit Delay', [model '/mu(z^-1)'], 'Position', [140 30 200 70]);
    add_block('simulink/Discrete/Unit Delay', [model '/sigma2(z^-1)'], 'Position', [140 90 200 130]);

    add_block('simulink/User-Defined Functions/MATLAB Function', [model '/KalmanUpdate'], ...
              'Position', [260 40 420 120]);
    set_param([model '/KalmanUpdate'], 'Script', ...
        ['function [mu_p, sigma2_p] = f(mu, sigma2, z, R)\n' ...
         '% 1D Kalman update: K = sigma2 / (sigma2 + R)\n' ...
         'S = sigma2 + R;\n' ...
         'K = sigma2 / S;\n' ...
         'mu_p = mu + K*(z - mu);\n' ...
         'sigma2_p = (1 - K)*sigma2;\n' ...
         'end\n']);

    add_block('simulink/Sinks/Out1', [model '/mu_plus'], 'Position', [470 45 500 65]);
    add_block('simulink/Sinks/Out1', [model '/sigma2_plus'], 'Position', [470 95 500 115]);

    add_line(model, 'z_meas/1', 'KalmanUpdate/3');
    add_line(model, 'R/1', 'KalmanUpdate/4');
    add_line(model, 'mu(z^-1)/1', 'KalmanUpdate/1');
    add_line(model, 'sigma2(z^-1)/1', 'KalmanUpdate/2');

    add_line(model, 'KalmanUpdate/1', 'mu_plus/1');
    add_line(model, 'KalmanUpdate/2', 'sigma2_plus/1');

    add_line(model, 'KalmanUpdate/1', 'mu(z^-1)/1', 'autorouting', 'on');
    add_line(model, 'KalmanUpdate/2', 'sigma2(z^-1)/1', 'autorouting', 'on');

    save_system(model);
    disp(['Saved Simulink model: ' model '.slx']);
end

File: Autonomous_Mobile_Robots_Course/Chapter9/Lesson4/Chapter9_Lesson4.py
#!/usr/bin/env python3
# Chapter9_Lesson4.py
# Elevation + traversability maps (outdoor AMR) — minimal, from-scratch implementation.

import math
import numpy as np
import matplotlib.pyplot as plt


def sigmoid(x: np.ndarray) -> np.ndarray:
    return 1.0 / (1.0 + np.exp(-x))


class ElevationTraversabilityMap:
    """
    Grid-based elevation map where each cell stores:
      - mu: posterior mean of terrain elevation (meters)
      - sigma2: posterior variance (m^2)

    Updates are done with a 1D Kalman filter per cell:
      z = h + v,  v ~ N(0, R)
    """

    def __init__(self, x_min, x_max, y_min, y_max, resolution,
                 sigma0=2.0,  # initial std dev in meters
                 meas_sigma=0.10,  # sensor std dev in meters
                 gate_k=3.0):
        self.x_min = float(x_min)
        self.x_max = float(x_max)
        self.y_min = float(y_min)
        self.y_max = float(y_max)
        self.res = float(resolution)

        self.nx = int(math.ceil((self.x_max - self.x_min) / self.res))
        self.ny = int(math.ceil((self.y_max - self.y_min) / self.res))

        # Mean and variance fields
        self.mu = np.zeros((self.nx, self.ny), dtype=np.float64)
        self.sigma2 = (sigma0 ** 2) * np.ones((self.nx, self.ny), dtype=np.float64)

        # Simple bookkeeping: number of fusions per cell
        self.count = np.zeros((self.nx, self.ny), dtype=np.int32)

        self.R = float(meas_sigma) ** 2
        self.gate_k = float(gate_k)

    def _to_index(self, x, y):
        ix = int((x - self.x_min) / self.res)
        iy = int((y - self.y_min) / self.res)
        if 0 <= ix < self.nx and 0 <= iy < self.ny:
            return ix, iy
        return None

    def update_point(self, x, y, z):
        idx = self._to_index(x, y)
        if idx is None:
            return False
        ix, iy = idx

        mu = self.mu[ix, iy]
        s2 = self.sigma2[ix, iy]

        # Innovation
        nu = z - mu
        S = s2 + self.R

        # Robust gate: if far from current belief, reject as obstacle/outlier
        if self.count[ix, iy] > 0 and abs(nu) > self.gate_k * math.sqrt(S):
            return False

        # Kalman gain
        K = s2 / S

        # Posterior
        self.mu[ix, iy] = mu + K * nu
        self.sigma2[ix, iy] = (1.0 - K) * s2
        self.count[ix, iy] += 1
        return True

    def update_point_cloud(self, pts_xyz):
        accepted = 0
        for x, y, z in pts_xyz:
            accepted += int(self.update_point(float(x), float(y), float(z)))
        return accepted

    def compute_features(self):
        """
        Compute local geometric features from elevation:
          - slope: arctan(||grad h||) (radians)
          - roughness: std-dev of neighborhood heights (meters)
          - step: max neighbor height difference (meters)
        """
        mu = self.mu
        nx, ny = mu.shape

        slope = np.zeros_like(mu)
        rough = np.zeros_like(mu)
        step = np.zeros_like(mu)

        # Use finite-difference gradients (skip borders)
        for i in range(1, nx - 1):
            for j in range(1, ny - 1):
                if self.count[i, j] == 0:
                    continue

                dzdx = (mu[i + 1, j] - mu[i - 1, j]) / (2.0 * self.res)
                dzdy = (mu[i, j + 1] - mu[i, j - 1]) / (2.0 * self.res)
                slope[i, j] = math.atan(math.sqrt(dzdx * dzdx + dzdy * dzdy))

                neigh = mu[i - 1:i + 2, j - 1:j + 2].reshape(-1)
                rough[i, j] = float(np.std(neigh))

                # Step: max absolute neighbor difference
                diffs = np.abs(neigh - mu[i, j])
                step[i, j] = float(np.max(diffs))

        return slope, rough, step

    def traversability_cost(self, slope, rough, step,
                            slope_ref=0.35,   # ~20 deg in rad
                            rough_ref=0.08,   # meters
                            step_ref=0.12,    # meters
                            w_s=3.0, w_r=2.0, w_d=2.5,
                            bias=-3.0):
        """
        Map (slope, roughness, step) -> [0,1] cost where 0 is easy and 1 is hard.
        """
        # Normalize features
        s = slope / max(1e-9, float(slope_ref))
        r = rough / max(1e-9, float(rough_ref))
        d = step / max(1e-9, float(step_ref))

        score = w_s * s + w_r * r + w_d * d + bias
        return sigmoid(score)

    def save_csv(self, path_prefix):
        np.savetxt(path_prefix + "_elevation_mu.csv", self.mu, delimiter=",")
        np.savetxt(path_prefix + "_elevation_sigma2.csv", self.sigma2, delimiter=",")
        np.savetxt(path_prefix + "_count.csv", self.count, delimiter=",")


def generate_synthetic_point_cloud(n_ground=50000, n_outliers=2000, seed=7):
    """
    Synthetic terrain z_true(x,y) + noisy measurements, plus some obstacle outliers.
    """
    rng = np.random.default_rng(seed)

    # Terrain domain
    x = rng.uniform(-10.0, 10.0, size=n_ground)
    y = rng.uniform(-10.0, 10.0, size=n_ground)

    # Smooth terrain
    z_true = (
        0.20 * np.sin(0.35 * x) +
        0.15 * np.cos(0.25 * y) +
        0.03 * x
    )

    # Add a hill/bump (still traversable)
    z_true += 0.25 * np.exp(-0.08 * ((x - 2.0) ** 2 + (y + 1.0) ** 2))

    # Sensor noise
    z_meas = z_true + rng.normal(0.0, 0.10, size=n_ground)

    ground = np.stack([x, y, z_meas], axis=1)

    # Outliers (e.g., vegetation/rocks) that should be rejected by gating
    xo = rng.uniform(-10.0, 10.0, size=n_outliers)
    yo = rng.uniform(-10.0, 10.0, size=n_outliers)
    zo = (
        0.20 * np.sin(0.35 * xo) +
        0.15 * np.cos(0.25 * yo) +
        0.03 * xo +
        0.8 + rng.normal(0.0, 0.05, size=n_outliers)
    )
    outliers = np.stack([xo, yo, zo], axis=1)

    pts = np.concatenate([ground, outliers], axis=0)
    rng.shuffle(pts, axis=0)
    return pts


def main():
    # Build map
    emap = ElevationTraversabilityMap(
        x_min=-10.0, x_max=10.0, y_min=-10.0, y_max=10.0,
        resolution=0.20,
        sigma0=2.0, meas_sigma=0.10, gate_k=3.0
    )

    pts = generate_synthetic_point_cloud()
    accepted = emap.update_point_cloud(pts)
    print("Points accepted:", accepted, "out of", pts.shape[0])

    slope, rough, step = emap.compute_features()
    cost = emap.traversability_cost(slope, rough, step)

    # Visualize
    plt.figure()
    plt.title("Elevation mean (m)")
    plt.imshow(emap.mu.T, origin="lower", aspect="equal")
    plt.colorbar()

    plt.figure()
    plt.title("Traversability cost (0 easy, 1 hard)")
    plt.imshow(cost.T, origin="lower", aspect="equal")
    plt.colorbar()

    plt.show()

    # Export
    emap.save_csv("Chapter9_Lesson4_output")


if __name__ == "__main__":
    main()
